presto-main/src/main/java/com/facebook/presto/operator/VariableWidthBlockEncodingBuffers.java

```diff
@@ -0,0 +1,257 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator;
+
+import com.facebook.presto.spi.block.AbstractVariableWidthBlock;
+import com.google.common.annotations.VisibleForTesting;
+import io.airlift.slice.SliceOutput;
+import org.openjdk.jol.info.ClassLayout;
+
+import java.util.Arrays;
+
+import static com.facebook.presto.operator.UncheckedByteArrays.setIntUnchecked;
+import static com.facebook.presto.spi.MoreByteArrays.setBytes;
+import static io.airlift.slice.SizeOf.SIZE_OF_INT;
+import static io.airlift.slice.SizeOf.sizeOf;
+import static java.lang.Math.max;
+import static sun.misc.Unsafe.ARRAY_INT_INDEX_SCALE;
+
+public class VariableWidthBlockEncodingBuffers
+        extends BlockEncodingBuffers
+{
+    @VisibleForTesting
+    public static final int POSITION_SIZE = Integer.BYTES + Byte.BYTES;
+
+    private static final String NAME = "VARIABLE_WIDTH";
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(VariableWidthBlockEncodingBuffers.class).instanceSize();
+    private static final int ESTIMATED_ELEMENT_SIZE = 8;
+
+    // The buffer for the slice for all incoming blocks so far
+    private byte[] sliceBuffer;
+
+    // The address that the next slice will be written to.
+    private int sliceBufferIndex;
+
+    // The buffer for the offsets for all incoming blocks so far
+    private byte[] offsetsBuffer;
+
+    // The address that the next offset value will be written to.
+    private int offsetsBufferIndex;
+
+    // The last offset in the offsets buffer
+    private int lastOffset;
+
+    public VariableWidthBlockEncodingBuffers(int initialPositionCount)
+    {
+        super(initialPositionCount);
+    }
+
+    @Override
+    public void resetBuffers()
+    {
+        bufferedPositionCount = 0;
+        sliceBufferIndex = 0;
+        offsetsBufferIndex = 0;
+        lastOffset = 0;
+        resetNullsBuffer();
+    }
+
+    @Override
+    public void accumulateRowSizes(int[] rowSizes)
+    {
+        int[] positions = getPositions();
+        for (int i = 0; i < positionCount; i++) {
+            rowSizes[i] += decodedBlock.getSliceLength(positions[i]);
```

Comment:
any reason not to combine these two statements: `rowSizes[i] += POSITION_SIZE + decodedBlock.getSliceLength(positions[i]);`